core/src/kura.rs

```diff
@@ -115,9 +118,41 @@ impl Kura {
     /// - file storage is unavailable
     /// - data in file storage is invalid or corrupted
     #[iroha_logger::log(skip_all, name = "kura_init")]
-    pub fn init(&self) -> Result<BlockCount> {
+    pub fn init(self: &Arc<Self>) -> Result<(Option<WorldStateView>, BlockCount)> {
         let block_store = self.block_store.lock();
 
+        let mut wsv = None;
+        let mut skip_block_count = 0;
+
+        if matches!(self.mode, Mode::Fast) {
+            let mut data = Vec::new();
+            match block_store.read_snapshot_data(&mut data) {
+                Ok(_) => {
+                    let seed = WsvSeed { kura: self.clone() };
+                    let mut deserializer = serde_json::Deserializer::from_slice(&data);
+                    match seed.deserialize(&mut deserializer) {
+                        Ok(read_wsv) => {
+                            skip_block_count = read_wsv.height();
+                            iroha_logger::trace!(?skip_block_count, "Read WSV from snapshot");
+                            wsv = Some(read_wsv);
+                        }
+                        Err(e) => {
+                            iroha_logger::warn!(
```

Comment:
In principle, yes. In practice, because it's a closure, therefore classified as a function, it changes the scope. Unless you know that this is inside a `span`, I'd be careful with using `and_then` and other monadics for logging.